src/build/sw-precache.ts

```diff
@@ -10,40 +10,161 @@
-    staticFileGlobs: precacheList,
-    navigateFallback: mainHtml,
-    //TODO: pull from user precache config
-    navigateFallbackWhitelist: [/\/data\/.*json/],
-    stripPrefix: buildRoot,
-    // verbose: true,
-  });
+  console.log('precacheList' , precacheList);
+
+  swConfig.stripPrefix = buildRoot;
+  swConfig.staticFileGlobs = precacheList;
+  // TODO: determine if this is a good default, or pull from user precache config
+  swConfig.navigateFallback = mainHtml;
+
+  return swPrecache.generate(swConfig);
+}
+
+export interface SWPreCacheTransformOptions {
+  /**
+   * folder containing files to be served by the service worker.
+   */
+  root: string;
+  /**
+   * Main file to serve from service worker.
+   */
+  main: string;
+  /**
+   * Output folder for the service worker bundle
+   */
+  buildRoot: string;
+  /**
+   * Name of the output service worker file.
+   */
+  serviceWorkerName: string;
+  /**
+   * Promise that returns the list of files to be cached by the service worker.
+   *
+   * If not given, all files streamed to SWPreCacheTransform will be put into the precache list, except for the file matching `configFileName`.
+   */
+  deps?: Promise<string[]>;
+  /**
+   * Existing config file to use as a base for the serivce worker generation.
+   *
+   * This file will not be copied into the output bundles.
+   */
+  configFileName?: string;
+}
+
+export class SWPreCacheTransform extends Transform {
+  swConfig: SWConfig;
+  options: SWPreCacheTransformOptions;
+  fileSet: Set<string>;
+  fullConfigFilePath: string;
+
+  constructor(options: SWPreCacheTransformOptions) {
+    super({objectMode: true});
+    this.options = options;
+    // if no given deps, collect all input files as deps
+    if (!options.deps) {
+      this.fileSet = new Set<string>();
+    }
+    if (options.configFileName) {
+      this.fullConfigFilePath = path.resolve(
+        options.root,
+        options.configFileName
+      );
+    }
+  }
+
+  _transform(file: File, encoding: string, callback:(error?, data?) => void): void {
+    if (file.path === this.fullConfigFilePath) {
+      this.swConfig = JSON.parse(file.contents.toString());
+      callback(null, null);
+    } else {
+      if (this.fileSet) {
+        this.fileSet.add(file.path);
+      }
+      callback(null, file);
+    }
+  }
+
+  end() {
+    let promise: Promise<string[]>;
+    if (this.fileSet) {
+      promise = Promise.resolve(Array.from(this.fileSet));
+    } else {
+      promise = this.options.deps;
+    }
+    promise.then((deps) => {
+      return generateServiceWorker(
+        this.options.root,
+        this.options.main,
+        deps,
+        this.options.buildRoot,
+        this.swConfig
+      );
+    })
+    .then((config) => {
+      let file = new File({
+        path: path.resolve(
+          this.options.buildRoot,
+          this.options.serviceWorkerName
+        ),
+        contents: new Buffer(config)
+      });
+      this.push(file);
+      super.end();
```

Comment:
is it cool to defer `end()` like this? I suspect it is, but deferring a super call is a pretty tricky thing to do generally...